### PR TITLE
Enable Development and Production in Kubernetes

### LINKED ISSUE #126: No Kubernetes Support
It would be awesome if development would not require Docker specifically, but rather Kubernetes (or any CRI). I'll take a look at implementing a minimal Kubernetes development environment using [Skaffold](https://skaffold.dev) as soon as possible if everyone here consents to it.

### PR DESCRIPTION
**Pull request by pojntfx**
Closes #126.

### CODE REVIEW

**Comment by pojntfx** [comment]
@dessalines This is currently using my [personal container registry](https://gitlab.com/pojntfx/lemmy/container_registry) for the images. Moving forward having an actual container registry in place would probably be a good idea. I see the following solutions to this:

- Drop the proprietary Travis CI in favor of the FLOSS GitLab CI/CD (would integrate nicely with Kubernetes as well; Skaffold could also be used for testing, CI/CD, you could run CI/CD on your local machine, GitHub integrates nicely with GitLab and so on and so forth)
- Use only GitLab's container registry and continue using Travis
- Use Docker Hub as the container registry and continue using Travis

I'd love to hear some thoughs on this.

**Comment by dessalines** [comment]
I'm trying to keep it *mostly* independent of any integrated build environment... I might use travis to set up automatic deploys for the main instance at some point, but really I'm just making sure to run `cargo test` before merging to master. That's really all travis is doing anyway. I'd like to avoid the work of setting up gitlab CI, and even tho I do have the repo pushing over there, I kinda want it to stay independent of it.

CI is nothing more than `cargo test` right now, I don't see the need to set up anything more complicated.

I've had a request to push these images to dockerhub, but its too much data for me to handle rn, the images are ~700MB and I live in the US with its data cap of 1TB / month.

Not to say I disagree with any of your solutions... the Gitlab CI one would be fine, I've just already done the work of setting up the travis one and don't wanna do it myself.

I'll review ur PR now.

File: README.md [hunk 1]
```diff
@@ -11,15 +11,16 @@
 
 A link aggregator / reddit clone for the fediverse.
 
-[Lemmy Dev instance](https://dev.lemmy.ml) *for testing purposes only*
+[Lemmy Dev instance](https://dev.lemmy.ml) _for testing purposes only_
```
Comment (dessalines) [inline README.md]: I prefer the * instead of _

File: README.md [hunk 2]
```diff
@@ -11,15 +11,16 @@
...
 
 This is a **very early beta version**, and a lot of features are currently broken or in active development, such as federation.
- 
-Front Page|Post
----|---
-![main screen](https://i.imgur.com/y64BtXC.png)|![chat screen](https://i.imgur.com/vsOr87q.png)
+
+| Front Page                                      | Post                                            |
+| ----------------------------------------------- | ----------------------------------------------- |
```
Comment (dessalines) [inline README.md]: I'm not a fan of these auto-width fillers, I'm guessing they were inserted by something automatically. We all don't use editors that have that same plugin so simpler is better.

File: README.md [hunk 3]
```diff
@@ -11,15 +11,16 @@
...
+| ![main screen](https://i.imgur.com/y64BtXC.png) | ![chat screen](https://i.imgur.com/vsOr87q.png) |
 
 ## Features
+
```
Comment (dessalines) [inline README.md]: These aren't necessary

File: README.md [hunk 4]
```diff
@@ -45,32 +47,75 @@ The overall goal is to create an easily self-hostable, decentralized alternative
 Each lemmy server can set its own moderation policy; appointing site-wide admins, and community moderators to keep out the trolls, and foster a healthy, non-toxic environment where all can feel comfortable contributing.
 
 ## Why's it called Lemmy?
+
 - Lead singer from [motorhead](https://invidio.us/watch?v=pWB5JZRGl0U).
-- The old school [video game](https://en.wikipedia.org/wiki/Lemmings_(video_game)).
+- The old school [video game](<https://en.wikipedia.org/wiki/Lemmings_(video_game)>).
```
Comment (dessalines) [inline README.md]: Hrm I've never seen this before! nice.

File: README.md [hunk 5]
```diff
@@ -45,32 +47,75 @@ The overall goal is to create an easily self-hostable, decentralized alternative
...
 git clone https://github.com/dessalines/lemmy
 cd lemmy
 ./docker_update.sh # This pulls the newest version, builds and runs it
 ```
+
 and goto http://localhost:8536
-### Local Development
-#### Requirements
+
+<!-- #### Kubernetes (WIP)
```
Comment (dessalines) [inline README.md]: We can just wait to merge all this until you're ready for it.

File: ui/stack.dev.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+kind: Service
+metadata:
+  name: lemmy-ui--dev
+spec:
+  type: NodePort
+  selector:
+    app: lemmy-ui--dev
+  ports:
+    - port: 4444
+      nodePort: 30002
```
Comment (dessalines) [inline ui/stack.dev.yaml]: Looks pretty neat, so this is used for development too? interesting.

Comment (pojntfx) [reply ui/stack.dev.yaml]: Yes, this is being used for everything, the whole workflow. I'm still implementing the production version though, currently this is dev only (port `4444` gets forwarded to localhost and hot-reloading works which makes this a very viable alternative to "native" development).

File: README.md (hunk 5 above)
Comment (pojntfx) [reply README.md]: That's probably a good idea, a production version should exist beforehand.

File: README.md (hunk 3 above)
Comment (pojntfx) [reply README.md]: I agree, these have just been automatically added by my formatter (`prettier`). Should I remove them again?

File: README.md (hunk 2 above)
Comment (pojntfx) [reply README.md]: Sure, I'll remove them. These have been automatically added by my formatter so kind of just them.

File: README.md (hunk 1 above)
Comment (pojntfx) [reply README.md]: Sure, I'll revert them back to the old ones. Just my formatter being overly aggressive.

**Comment by pojntfx** [comment]
> I'm trying to keep it _mostly_ independent of any integrated build environment... I might use travis to set up automatic deploys for the main instance at some point, but really I'm just making sure to run `cargo test` before merging to master. That's really all travis is doing anyway. I'd like to avoid the work of setting up gitlab CI, and even tho I do have the repo pushing over there, I kinda want it to stay independent of it.
> 
> CI is nothing more than `cargo test` right now, I don't see the need to set up anything more complicated.
> 
> I've had a request to push these images to dockerhub, but its too much data for me to handle rn, the images are ~700MB and I live in the US with its data cap of 1TB / month.
> 
> Not to say I disagree with any of your solutions... the Gitlab CI one would be fine, I've just already done the work of setting up the travis one and don't wanna do it myself.
> 
> I'll review ur PR now.

Where I live, a local cooperative is the ISP so I could push the images if you want to. Alternatively, using GitLab's CI it would also be possible to let the runner push the images. I might take a look at setting up a CI/CD configuration as it is very little work now that Skaffold is being used. Thanks for the review, I'll take a look a the issues ASAP.

**Comment by pojntfx** [comment]
@dessalines A production version has been added and the development workflow now works fluently. I've also added documentation. Would you mind to do a second review?

File: README.md (hunk 3 above)
Comment (dessalines) [reply README.md]: Yes plz.

**Comment by dessalines** [comment]
I'll merge this once the newlines are removed.

**Comment by pojntfx** [comment]
Alright, the issues have been addressed. Should be merge-able now!

**Pull request merged**